crates/shielder-relayer/src/price_feed.rs

```diff
@@ -0,0 +1,125 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use rust_decimal::Decimal;
+use serde::Deserialize;
+use strum::IntoEnumIterator as _;
+use strum_macros::EnumIter;
+use time::OffsetDateTime;
+use tokio::time::Duration;
+
+const BASE_PATH: &str = "https://api.diadata.org/v1/assetQuotation";
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, EnumIter)]
+pub enum Coin {
+    Eth,
+    Azero,
+    Btc,
+    Usdt,
+    Usdc,
+}
+
+/// A collection of prices for various coins.
+///
+/// The underlying structure is behind a mutex and a process to update it
+/// asynchronously can be strated with `start_price_feed`.
+#[derive(Clone)]
+pub struct Prices {
+    validity: time::Duration,
+    refresh_interval: Duration,
+    inner: Arc<Mutex<HashMap<Coin, Price>>>,
+}
+
+#[derive(Clone, Deserialize, Debug)]
+struct Price {
+    #[serde(rename = "Price")]
+    price: Decimal,
+    #[serde(
+        rename = "Time",
+        deserialize_with = "time::serde::iso8601::deserialize"
+    )]
+    time: OffsetDateTime,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Reqwest error: {0}")]
+    Reqwest(#[from] reqwest::Error),
+}
+
+impl Prices {
+    /// Create a new `Prices` instance with the given validity and refresh interval.
+    ///
+    /// Note that you should realistically set `validity` to at least 5 or 10 minutes - it seems
+    /// the API we are using (DIA) updates about 2 or 3 minutes or so.
+    pub fn new(validity: Duration, refresh_interval: Duration) -> Self {
+        let inner = Default::default();
+        let validity =
+            time::Duration::new(validity.as_secs() as i64, validity.subsec_nanos() as i32);
+
+        Self {
+            validity,
+            refresh_interval,
+            inner,
+        }
+    }
+
+    /// Get the price of a coin or `None` if the price is not available or outdated.
+    pub fn price(&self, coin: Coin) -> Option<Decimal> {
+        let inner = self.inner.lock().expect("Mutex poisoned");
```

Comment:
is there an obvious reason to work with synchronous locking? in general, does it make sense to make `PriceFeed` async?

Reply:
Hm, I just default to this. What is the rationale to prefer async? I took care so that the lock was taken out for just the time needed to get/update the map, so it shuoldn't ever block any tokio worker.

Reply:
to prefer async - everything else in relayer is async

Reply:
I don't trust the async mutexes, but sure, can change.